### ISSUE TITLE
[feature] Add the ability to dump all commands and parameters to a json file

### ISSUE THREAD

**Issue by KathleenDollard** [dotnet/command-line-api#401]
This was requested of the .NET CLI https://github.com/dotnet/cli/issues/10673. I feel it is needed across enough CLI's to warrant consideration as a feature of System.CommandLine. Key part is marked.

"Regarding the CLI docs published here: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet?tabs=netcore21

One thing that will help us get docs published easily for CLI reference is if we can _dump the whole list of commands, descriptions, option/mandatory status, parameters list, etc_. This way we can easily diff between releases to see what has changed. It will also allow us to automate the documentation!

It's very troublesome to update the CLI ref which is literally a copy/paste of output from running the commands by hand..."

**Comment by jonsequitur** [comment]
This is a great idea. I would imagine implementing it as a directive. What would we expect the output format(s) to look like? Plain text seems clear. Would Markdown be useful?

Does it relate to another thing we've discussed which is being able to serialize and deserialize a parser grammar? Both could be used for example for ApprovalTest-style testing.

**Comment by Thraka** [comment]
I don't think you should output to a specific "style" like markdown, html, plain text. It should be a structured  data file like .json. This way, the capabilities the cli reports can be consumed and transformed into whatever is needed for publishing docs. 

I first discovered that the azure x-plat cli (defunct now) would output a file like this by calling `azure.exe --output-help` or something like that. I wrote a tool that spun up a razor engine that would consume the output.json file and then spit out multiple individual .md files to doc each command.

**Comment by oising** [comment]
I was going to log a separate issue for my idea, but this one seems close enough - there really should be a common parameter to dump the parameters to _stdout_ as json. This would let powershell / powershell core's command line completion to query any binaries and provide tab completion very easily with a native completer. 

See this for more information.
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/register-argumentcompleter?view=powershell-6

If this could make your 1.0 milestone, you'd seriously aid in adoption if this library would enable rich tab completion for powershell/pwsh.

**Comment by jonsequitur** [comment]
For `dotnet-suggest` we have a generalized completer that works for any app using `System.CommandLine`: https://github.com/dotnet/command-line-api/wiki/dotnet-suggest. Both PowerShell and bash are currently supported.
